### ISSUE TITLE
Two-tower ranking model — train learned (user, post) scorer

### ISSUE THREAD

**Issue by Shaglees** [Shaglees/beepbopboop#42]
## Parent Epic
Part of #39 — ML-powered feed: vector embeddings and learned ranking

## Objective
Train a two-tower neural network that takes a (user embedding, post embedding) pair and outputs a relevance score. This learned scorer replaces/augments the hand-crafted `scorePost` function in `post_repo.go:481`.

## Background

The current `scorePost` function manually combines recency, geo proximity, label affinity, and type affinity using hand-tuned weights. A two-tower model learns these relationships from data — and discovers patterns (e.g. "users who save weather posts also engage with local events") that the hand-crafted function cannot capture.

### Two-Tower Architecture

```
User Tower          Post Tower
──────────          ──────────
user_embed (1536)   post_embed (1536)
     │                   │
   Linear(256)         Linear(256)
   ReLU               ReLU
   Linear(128)         Linear(128)
   L2 norm             L2 norm
     │                   │
  user_repr (128)    post_repr (128)
       └──────┬──────┘
              │
          dot product
              │
      + context_features
        (recency, geo_dist,
         hour_of_day, day_of_week)
              │
          Linear(32) → Linear(1)
              │
          relevance score
```

**Why two towers?**
- User tower runs once per request (precomputed)
- Post tower runs once per post (cached)
- Efficient: only the final MLP runs at inference per (user, post) pair
- Proven at scale: YouTube DNN (2016), TikTok, Pinterest

### Labels
- **Positive (1.0):** dwell_ms >= 10000ms OR save OR reaction='more'
- **Weak positive (0.7):** dwell_ms 3000–10000ms OR click
- **Implicit negative (0.0):** view with dwell_ms < 3000ms, no other interaction
- **Hard negative (-0.5 → 0.0):** reaction='less' OR reaction='not_for_me'

### Loss Function
Binary cross-entropy on relevance score. For hard negatives, use margin ranking loss to push negative pairs further from positive pairs.

### Training Framework
- Python training script (separate from Go backend): `ml/train.py`
- PyTorch or JAX (Karpathy's `micrograd`/`nanoGPT` style: small, readable, no magic)
- Export trained weights to ONNX for Go inference via `onnxruntime-go`
- Model size target: < 2MB (fits in memory, fast inference)

### Evaluation Metrics
- **AUC-ROC** on held-out engagement pairs (target > 0.75)
- **NDCG@10** — are the top 10 posts for a user the ones they actually engaged with?
- **Offline A/B proxy:** precision@k on saved posts

### 1. Project structure
```
ml/
  train.py              # Training loop
  model.py              # TwoTowerModel class
  dataset.py            # Dataset loading from training data exports
  evaluate.py           # Offline evaluation metrics
  export.py             # Export to ONNX
  requirements.txt      # torch, onnx, numpy, psycopg2
```

### 2. Model definition (`model.py`)
```python
class UserTower(nn.Module):
    def __init__(self, input_dim=1536, hidden=256, out_dim=128):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(input_dim, hidden), nn.ReLU(),
            nn.Linear(hidden, out_dim),
        )
    def forward(self, x):
        return F.normalize(self.net(x), dim=-1)

class TwoTowerModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.user_tower = UserTower()
        self.post_tower = UserTower()  # same arch, separate weights
        self.context_mlp = nn.Sequential(
            nn.Linear(128*2 + 4, 32), nn.ReLU(), nn.Linear(32, 1)
        )
```

### 3. Training loop
- Load training data from `ml/data/training_pairs.parquet` (produced by #42)
- 80/10/10 train/val/test split stratified by user
- Adam optimizer, LR=1e-3, cosine decay
- Early stopping on val AUC

### 4. ONNX export
```python
# export.py
torch.onnx.export(model, (user_vec, post_vec, ctx), "ranker.onnx",
    input_names=["user_vec","post_vec","ctx"],
    output_names=["score"])
```

### 5. Integration (Go side)
- Load `ranker.onnx` at server startup via `onnxruntime-go`
- Expose `Score(userVec, postVec, ctx []float32) float32` function
- Used in `multi_feed.go` ForYou ranking path

## Testing Requirements

1. Model trains to AUC > 0.70 on held-out data with 1000+ user-post pairs
2. ONNX export runs correctly and matches PyTorch output within float32 precision
3. Go inference returns score in < 5ms per pair
4. Model file < 5MB

## Success Criteria
- [ ] `ml/` directory with model, training, evaluation scripts
- [ ] Model trains successfully on exported training data
- [ ] Offline AUC-ROC > 0.75 on held-out test set
- [ ] ONNX export works; Go can load and run inference
- [ ] Integration with ForYou feed (alongside current scoring, not replacing it yet)

## Key Files
- `backend/internal/repository/post_repo.go:481` — `scorePost()` to augment/replace
- `backend/internal/handler/multi_feed.go` — ForYou feed integration point
- Related: #40 (post embeddings), #41 (user embeddings), #42 (training data)
